services/net/filemonitor/FilemonitorAction.cs

```diff
@@ -0,0 +1,678 @@
+                    item.Summary = GetFmsData(preFiltered + "<break>", Fields.Story, dataSource);
+                    item.Language = GetFmsData(filtered, Fields.Lang, dataSource);
+                    item.FilePath = document.Key;
+                    item.Abstract = GetFmsData(filtered, Fields.Summary, dataSource);
+                    item.Page = GetFmsData(filtered, Fields.Page, dataSource);
+                    item.Section = GetFmsData(filtered, Fields.Section, dataSource);
+
+                    item.Tags = GetTagList(filtered, dataSource);
+                    item.Authors = GetAuthorList(GetFmsData(filtered, Fields.Author, dataSource));
+                    item.PublishedOn = GetPublishedOn(GetFmsData(filtered, Fields.Date, dataSource), dataSource);
+
+                    articles.Add(item);
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogInformation(ex, $"File contents for data source '{dataSource.Code}' is invalid.");
+                throw;
+            }
+        }
+
+        return articles;
+    }
+
+    /// <summary>
+    /// Get the data source's "code" value. If the data source is selfPublished (relates to only one publication) the code
+    /// of the data source is returned. If the content string attribute "selfPublished" is false, the actual data source for
+    /// the publication is retrieved by the value of the paperName parameter.
+    ///
+    /// The name of each paper is not guaranteed to be unique, so we may have to define multiple data sources for the same
+    /// publication, one for each unique name.
+    /// </summary>
+    /// <param name="dataSource"></param>
+    /// <param name="paperName"></param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    private async Task<string> GetItemSourceCodeAsync(DataSourceModel dataSource, string paperName)
+    {
+        // The default for selfPublished is false, so all self-published papers must have a value in connection string.
+        var selfPublished = GetConnectionValue(dataSource, Fields.SelfPublished);
+        selfPublished = !String.IsNullOrEmpty(selfPublished) ? selfPublished : "true";
+
+        if (Boolean.Parse(selfPublished))
+        {
+            return dataSource.Code;
+        }
+        else
+        {
+            var childDataSource = await _api.GetDataSourceByNameAsync(paperName);
+            return childDataSource != null ? childDataSource.Code : "No code found for: " + paperName;
+        }
+    }
+
+    /// <summary>
+    /// Get a value from the data source Connection string. If the key does not exist, return the empty string.
+    /// </summary>
+    /// <param name="dataSource"></param>
+    /// <param name="key"></param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    private static string GetConnectionValue(DataSourceModel dataSource, string key)
```

Comment:
This was because I based the filemonitor service on the syndication service. My GetConnectionValue() is a modification of syndication's GetUrl method. I just happened to rename it to use the same name as the extension method. I'll modify the code to use the extension method instead.

Reply:
Fixed.